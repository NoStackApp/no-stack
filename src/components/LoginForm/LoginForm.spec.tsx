import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import faker from 'faker';

import { RawLoginForm as LoginForm } from '.';

faker.seed(123);

describe('<LoginForm />', () => {
  const loginMock = jest.fn();
  let loginExceptionMock: any;
  let submitButtonText: string;
  let username: string;
  let password: string;

  beforeEach(() => {
    loginMock.mockReset();
    submitButtonText = 'Log In';
    username = faker.random.word();
    password = faker.random.word();
  });

  it('should initially not render', () => {
    const { container } = render(<LoginForm loading login={loginMock} />);

    expect(container.firstChild).toBeNull();
  });

  it('should not render if logged in', () => {
    const { container } = render(
      <LoginForm loading={false} currentUser login={loginMock} />,
    );

    expect(container.firstChild).toBeNull();
  });

  it('should render login form if not signed in', () => {
    const { getByLabelText, getByText } = render(
      <LoginForm login={loginMock} submitButtonText={submitButtonText} />,
    );

    const usernameInput = getByLabelText(/username/i);
    const passwordInput = getByLabelText(/password/i);
    const button = getByText(submitButtonText);

    fireEvent.change(usernameInput, {
      target: {
        value: username,
      },
    });

    fireEvent.change(passwordInput, {
      target: {
        value: password,
      },
    });

    fireEvent.click(button);

    expect(loginMock).toHaveBeenCalledTimes(1);
    expect(loginMock).toHaveBeenCalledWith({
      username,
      password,
    });
  });

  describe('AND login fails', () => {
    beforeEach(() => {
      loginExceptionMock = jest.fn(() => {
        throw new Error('error');
      });
    });
    it('SHOULD set the error state', () => {
      const {
        getByRole,
        container: { firstChild },
      } = render(
        <LoginForm
          submitButtonText={submitButtonText}
          login={loginExceptionMock}
        />,
      );

      expect(firstChild).toMatchSnapshot();

      const form = getByRole('form');
      fireEvent.submit(form);
      expect(loginExceptionMock).toHaveBeenCalled();
    });
  });
});
