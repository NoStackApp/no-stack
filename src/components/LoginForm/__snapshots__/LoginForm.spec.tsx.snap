// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> AND login fails SHOULD set the error state 1`] = `
<div
  class="sc-bdVaJa ezJIro"
>
  <form>
    <div
      class="sc-bwzfXH hWAxya"
    >
      <label
        for="nostack-username"
      >
        Username:
        <input
          id="nostack-username"
          name="username"
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      class="sc-bwzfXH hWAxya"
    >
      <label
        for="nostack-password"
      >
        Password:
        <input
          id="nostack-password"
          name="password"
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      class="sc-bwzfXH hWAxya"
    >
      <button
        disabled=""
        type="submit"
      >
        Log In
      </button>
    </div>
    
  </form>
</div>
`;
